File: technology/kubernetes/demo-app/src/main/java/org/acme/demoapp/generator/DemoDataGenerator.java
package org.acme.demoapp.generator;

import java.time.DayOfWeek;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import jakarta.enterprise.context.ApplicationScoped;

import org.acme.common.domain.Lesson;
import org.acme.common.domain.Room;
import org.acme.common.domain.TimeTable;
import org.acme.common.domain.Timeslot;

@ApplicationScoped
public final class DemoDataGenerator {

    private static final String[] ROOM_NAMES = new String[] {
            "Room A", "Room B", "Room C", "Room D", "Room E", "Room F", "Room G", "Room H", "Room I", "Room J"
    };

    private static final String[] SUBJECTS = new String[] {
            "Math", "Physics", "Chemistry", "Biology", "History", "English", "Art", "Drama", "ICT", "Physical education",
            "French", "Spanish", "Geography", "Geology"
    };

    private static final String[] STUDENT_GROUPS = new String[] {
            "9th grade", "10th grade", "11th grade", "12th grade"
    };

    private Map<String, String> subjectToTeacherMap;

    public DemoDataGenerator() {
        this.subjectToTeacherMap = createSubjectsToTeachersMapping();
    }

    private Map<String, String> createSubjectsToTeachersMapping() {
        Map<String, String> mapping = new HashMap<>();
        mapping.put("ICT", "A. Turing");
        mapping.put("Math", "A. Turing");
        mapping.put("Chemistry", "M. Curie");
        mapping.put("French", "M. Curie");
        mapping.put("Physics", "M. Curie");
        mapping.put("Biology", "C. Darwin");
        mapping.put("Geography", "C. Darwin");
        mapping.put("Geology", "C. Darwin");
        mapping.put("Drama", "I. Jones");
        mapping.put("English", "I. Jones");
        mapping.put("History", "I. Jones");
        mapping.put("Art", "S. Dali");
        mapping.put("Physical education", "C. Lewis");
        mapping.put("Spanish", "P. Cruz");
        return mapping;
    }

    public TimeTable generateUnsolvedTimeTable(long problemId, int lessons, int rooms) {
        final int timeslotCount = (int) (lessons * 1.5 / rooms);
        List<Timeslot> timeslotList = generateTimeslots(problemId, timeslotCount);
        List<Lesson> lessonList = generateLessons(problemId, lessons);
        List<Room> roomList = generateRooms(problemId, rooms);
        return new TimeTable(timeslotList, roomList, lessonList);
    }

    List<Timeslot> generateTimeslots(long problemId, int count) {
        int sanitizedCount = count > 30 ? 30 : count; // Maximum of 30 timeslots per week.
        LocalTime startTime = LocalTime.of(8, 0);
        DayOfWeek startDay = DayOfWeek.MONDAY;
        DayOfWeek currentDay = startDay;
        LocalTime currentStartTime = startTime;
        List<Timeslot> timeslots = new ArrayList<>(sanitizedCount);
        for (int i = 0; i < sanitizedCount; i++) {
            Timeslot timeslot = new Timeslot(problemId, currentDay, currentStartTime, currentStartTime.plusHours(1));
            timeslots.add(timeslot);

            if (currentStartTime.getHour() == 10) { // Lunch break.
                currentStartTime = currentStartTime.plusHours(3);
            } else if (currentStartTime.getHour() == 15) { // The next day.
                currentStartTime = startTime;
                currentDay = currentDay.plus(1);
            } else {
                currentStartTime = currentStartTime.plusHours(1);
            }
        }

        return timeslots;
    }

    List<Room> generateRooms(long problemId, int count) {
        int sanitizedCount = count > ROOM_NAMES.length ? ROOM_NAMES.length : count;
        List<Room> rooms = new ArrayList<>(sanitizedCount);
        for (int i = 0; i < sanitizedCount; i++) {
            Room room = new Room(ROOM_NAMES[i]);
            room.setProblemId(problemId);
            rooms.add(room);
        }
        return rooms;
    }

    List<Lesson> generateLessons(long problemId, int count) {
        int sanitizedCount = count > 200 ? 200 : count;
        List<Lesson> lessons = new ArrayList<>(sanitizedCount);
        for (int i = 0; i < sanitizedCount; i++) {
            int subjectIndex = i % SUBJECTS.length;
            int studentGroupIndex = i % STUDENT_GROUPS.length;
            String subject = SUBJECTS[subjectIndex];
            String studentGroup = STUDENT_GROUPS[studentGroupIndex];
            Lesson lesson = new Lesson(problemId, subject, subjectToTeacherMap.get(subject), studentGroup);
            lessons.add(lesson);
        }
        return lessons;
    }
}

File: technology/kotlin-quarkus/src/main/kotlin/org/acme/kotlin/schooltimetabling/bootstrap/DemoDataGenerator.kt
package org.acme.kotlin.schooltimetabling.bootstrap

import io.quarkus.runtime.StartupEvent
import org.acme.kotlin.schooltimetabling.domain.Lesson
import org.acme.kotlin.schooltimetabling.domain.Room
import org.acme.kotlin.schooltimetabling.domain.Timeslot
import org.acme.kotlin.schooltimetabling.persistence.LessonRepository
import org.acme.kotlin.schooltimetabling.persistence.RoomRepository
import org.acme.kotlin.schooltimetabling.persistence.TimeslotRepository
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.time.DayOfWeek
import java.time.LocalTime
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.event.Observes
import jakarta.inject.Inject
import jakarta.transaction.Transactional


@ApplicationScoped
class DemoDataGenerator {

    @ConfigProperty(name = "timeTable.demoData", defaultValue = "SMALL")
    lateinit var demoData: DemoData

    @Inject
    lateinit var timeslotRepository: TimeslotRepository
    @Inject
    lateinit var roomRepository: RoomRepository
    @Inject
    lateinit var lessonRepository: LessonRepository

    @Transactional
    fun generateDemoData(@Observes startupEvent: StartupEvent) {
        if (demoData == DemoData.NONE) {
            return
        }

        val timeslotList: MutableList<Timeslot> = mutableListOf(
                Timeslot(DayOfWeek.MONDAY, LocalTime.of(8, 30), LocalTime.of(9, 30)),
                Timeslot(DayOfWeek.MONDAY, LocalTime.of(9, 30), LocalTime.of(10, 30)),
                Timeslot(DayOfWeek.MONDAY, LocalTime.of(10, 30), LocalTime.of(11, 30)),
                Timeslot(DayOfWeek.MONDAY, LocalTime.of(13, 30), LocalTime.of(14, 30)),
                Timeslot(DayOfWeek.MONDAY, LocalTime.of(14, 30), LocalTime.of(15, 30)),

                Timeslot(DayOfWeek.TUESDAY, LocalTime.of(8, 30), LocalTime.of(9, 30)),
                Timeslot(DayOfWeek.TUESDAY, LocalTime.of(9, 30), LocalTime.of(10, 30)),
                Timeslot(DayOfWeek.TUESDAY, LocalTime.of(10, 30), LocalTime.of(11, 30)),
                Timeslot(DayOfWeek.TUESDAY, LocalTime.of(13, 30), LocalTime.of(14, 30)),
                Timeslot(DayOfWeek.TUESDAY, LocalTime.of(14, 30), LocalTime.of(15, 30)))

        if (demoData == DemoData.LARGE) {
            timeslotList.addAll(listOf(
                    Timeslot(DayOfWeek.WEDNESDAY, LocalTime.of(8, 30), LocalTime.of(9, 30)),
                    Timeslot(DayOfWeek.WEDNESDAY, LocalTime.of(9, 30), LocalTime.of(10, 30)),
                    Timeslot(DayOfWeek.WEDNESDAY, LocalTime.of(10, 30), LocalTime.of(11, 30)),
                    Timeslot(DayOfWeek.WEDNESDAY, LocalTime.of(13, 30), LocalTime.of(14, 30)),
                    Timeslot(DayOfWeek.WEDNESDAY, LocalTime.of(14, 30), LocalTime.of(15, 30)),
                    Timeslot(DayOfWeek.THURSDAY, LocalTime.of(8, 30), LocalTime.of(9, 30)),
                    Timeslot(DayOfWeek.THURSDAY, LocalTime.of(9, 30), LocalTime.of(10, 30)),
                    Timeslot(DayOfWeek.THURSDAY, LocalTime.of(10, 30), LocalTime.of(11, 30)),
                    Timeslot(DayOfWeek.THURSDAY, LocalTime.of(13, 30), LocalTime.of(14, 30)),
                    Timeslot(DayOfWeek.THURSDAY, LocalTime.of(14, 30), LocalTime.of(15, 30)),
                    Timeslot(DayOfWeek.FRIDAY, LocalTime.of(8, 30), LocalTime.of(9, 30)),
                    Timeslot(DayOfWeek.FRIDAY, LocalTime.of(9, 30), LocalTime.of(10, 30)),
                    Timeslot(DayOfWeek.FRIDAY, LocalTime.of(10, 30), LocalTime.of(11, 30)),
                    Timeslot(DayOfWeek.FRIDAY, LocalTime.of(13, 30), LocalTime.of(14, 30)),
                    Timeslot(DayOfWeek.FRIDAY, LocalTime.of(14, 30), LocalTime.of(15, 30))))
        }
        timeslotRepository.persist(timeslotList)


        val roomList: MutableList<Room> = mutableListOf(
                Room("Room A"),
                Room("Room B"),
                Room("Room C"))
        if (demoData == DemoData.LARGE) {
            roomList.addAll(listOf(
                    Room("Room D"),
                    Room("Room E"),
                    Room("Room F")))
        }
        roomRepository.persist(roomList)


        val lessonList: MutableList<Lesson> = mutableListOf(
                Lesson("Math", "A. Turing", "9th grade"),
                Lesson("Math", "A. Turing", "9th grade"),
                Lesson("Physics", "M. Curie", "9th grade"),
                Lesson("Chemistry", "M. Curie", "9th grade"),
                Lesson("Biology", "C. Darwin", "9th grade"),
                Lesson("History", "I. Jones", "9th grade"),
                Lesson("English", "I. Jones", "9th grade"),
                Lesson("English", "I. Jones", "9th grade"),
                Lesson("Spanish", "P. Cruz", "9th grade"),
                Lesson("Spanish", "P. Cruz", "9th grade"))
        if (demoData == DemoData.LARGE) {
            lessonList.addAll(listOf(
                    Lesson("Math", "A. Turing", "9th grade"),
                    Lesson("Math", "A. Turing", "9th grade"),
                    Lesson("Math", "A. Turing", "9th grade"),
                    Lesson("ICT", "A. Turing", "9th grade"),
                    Lesson("Physics", "M. Curie", "9th grade"),
                    Lesson("Geography", "C. Darwin", "9th grade"),
                    Lesson("Geology", "C. Darwin", "9th grade"),
                    Lesson("History", "I. Jones", "9th grade"),
                    Lesson("English", "I. Jones", "9th grade"),
                    Lesson("Drama", "I. Jones", "9th grade"),
                    Lesson("Art", "S. Dali", "9th grade"),
                    Lesson("Art", "S. Dali", "9th grade"),
                    Lesson("Physical education", "C. Lewis", "9th grade"),
                    Lesson("Physical education", "C. Lewis", "9th grade"),
                    Lesson("Physical education", "C. Lewis", "9th grade")))
        }

        lessonList.addAll(listOf(
                Lesson("Math", "A. Turing", "10th grade"),
                Lesson("Math", "A. Turing", "10th grade"),
                Lesson("Math", "A. Turing", "10th grade"),
                Lesson("Physics", "M. Curie", "10th grade"),
                Lesson("Chemistry", "M. Curie", "10th grade"),
                Lesson("French", "M. Curie", "10th grade"),
                Lesson("Geography", "C. Darwin", "10th grade"),
                Lesson("History", "I. Jones", "10th grade"),
                Lesson("English", "P. Cruz", "10th grade"),
                Lesson("Spanish", "P. Cruz", "10th grade")))
        if (demoData == DemoData.LARGE) {
            lessonList.addAll(listOf(
                    Lesson("Math", "A. Turing", "10th grade"),
                    Lesson("Math", "A. Turing", "10th grade"),
                    Lesson("ICT", "A. Turing", "10th grade"),
                    Lesson("Physics", "M. Curie", "10th grade"),
                    Lesson("Biology", "C. Darwin", "10th grade"),
                    Lesson("Geology", "C. Darwin", "10th grade"),
                    Lesson("History", "I. Jones", "10th grade"),
                    Lesson("English", "P. Cruz", "10th grade"),
                    Lesson("English", "P. Cruz", "10th grade"),
                    Lesson("Drama", "I. Jones", "10th grade"),
                    Lesson("Art", "S. Dali", "10th grade"),
                    Lesson("Art", "S. Dali", "10th grade"),
                    Lesson("Physical education", "C. Lewis", "10th grade"),
                    Lesson("Physical education", "C. Lewis", "10th grade"),
                    Lesson("Physical education", "C. Lewis", "10th grade"),

                    Lesson("Math", "A. Turing", "11th grade"),
                    Lesson("Math", "A. Turing", "11th grade"),
                    Lesson("Math", "A. Turing", "11th grade"),
                    Lesson("Math", "A. Turing", "11th grade"),
                    Lesson("Math", "A. Turing", "11th grade"),
                    Lesson("ICT", "A. Turing", "11th grade"),
                    Lesson("Physics", "M. Curie", "11th grade"),
                    Lesson("Chemistry", "M. Curie", "11th grade"),
                    Lesson("French", "M. Curie", "11th grade"),
                    Lesson("Physics", "M. Curie", "11th grade"),
                    Lesson("Geography", "C. Darwin", "11th grade"),
                    Lesson("Biology", "C. Darwin", "11th grade"),
                    Lesson("Geology", "C. Darwin", "11th grade"),
                    Lesson("History", "I. Jones", "11th grade"),
                    Lesson("History", "I. Jones", "11th grade"),
                    Lesson("English", "P. Cruz", "11th grade"),
                    Lesson("English", "P. Cruz", "11th grade"),
                    Lesson("English", "P. Cruz", "11th grade"),
                    Lesson("Spanish", "P. Cruz", "11th grade"),
                    Lesson("Drama", "P. Cruz", "11th grade"),
                    Lesson("Art", "S. Dali", "11th grade"),
                    Lesson("Art", "S. Dali", "11th grade"),
                    Lesson("Physical education", "C. Lewis", "11th grade"),
                    Lesson("Physical education", "C. Lewis", "11th grade"),
                    Lesson("Physical education", "C. Lewis", "11th grade"),

                    Lesson("Math", "A. Turing", "12th grade"),
                    Lesson("Math", "A. Turing", "12th grade"),
                    Lesson("Math", "A. Turing", "12th grade"),
                    Lesson("Math", "A. Turing", "12th grade"),
                    Lesson("Math", "A. Turing", "12th grade"),
                    Lesson("ICT", "A. Turing", "12th grade"),
                    Lesson("Physics", "M. Curie", "12th grade"),
                    Lesson("Chemistry", "M. Curie", "12th grade"),
                    Lesson("French", "M. Curie", "12th grade"),
                    Lesson("Physics", "M. Curie", "12th grade"),
                    Lesson("Geography", "C. Darwin", "12th grade"),
                    Lesson("Biology", "C. Darwin", "12th grade"),
                    Lesson("Geology", "C. Darwin", "12th grade"),
                    Lesson("History", "I. Jones", "12th grade"),
                    Lesson("History", "I. Jones", "12th grade"),
                    Lesson("English", "P. Cruz", "12th grade"),
                    Lesson("English", "P. Cruz", "12th grade"),
                    Lesson("English", "P. Cruz", "12th grade"),
                    Lesson("Spanish", "P. Cruz", "12th grade"),
                    Lesson("Drama", "P. Cruz", "12th grade"),
                    Lesson("Art", "S. Dali", "12th grade"),
                    Lesson("Art", "S. Dali", "12th grade"),
                    Lesson("Physical education", "C. Lewis", "12th grade"),
                    Lesson("Physical education", "C. Lewis", "12th grade"),
                    Lesson("Physical education", "C. Lewis", "12th grade")))
        }

        val lesson: Lesson = lessonList[0]
        lesson.timeslot = timeslotList[0]
        lesson.room = roomList[0]

        lessonRepository.persist(lessonList)
    }

    enum class DemoData {
        NONE, SMALL, LARGE
    }

}
